### ISSUE TITLE
Pipelined engine cannot be constructed for bundles whose logits output has a static sequence length (GrowingLogitsBuffer initialCapacity 256 vs [1, 1, vocab])

### ISSUE THREAD

**Issue by john-rocky** [apple/coreai-models#212]
Thanks for #103 — this is the same class of problem in the language-model path, and the same fix shape applies.

**Symptom.** `EngineFactory.createEngine` throws `Shape at dimension 1 of 256 is not a valid substitution for source shape 1` for any bundle whose `main` graph emits logits as a static `[1, 1, vocab]` (decode-only, S=1 graphs). It fails before any token is bound, so the prompt length does not matter. A third-party harness hit it on a public bundle: Liquid4All/pipette-clients#13 (`mlboydaisuke/Qwen3.8-27B-CoreAI`).

**Cause.** `CoreAIPipelinedEngine.swift:745-752` builds the logits buffer with `initialCapacity: averageExpectedPromptSize` (256, line 29), and `TensorStorage+CoreAI.swift:58-59` resolves `[1, capacity, vocabSize]` against the descriptor; a static dim-1 of 1 rejects 256.

**Fix (5 lines).** Size the buffer to the static length when the descriptor has one. This has been in production in the fork since June (john-rocky/coreai-models@9e5b605) on the community bundles published at huggingface.co/mlboydaisuke:

```swift
// A fully static logits output (e.g. a decode-only S=1 graph: [1, 1, vocab])
// can't be resolved at a larger capacity — size the buffer to its static
// sequence length instead of the prompt-sized default.
let logitsSeqIsStatic = logitsDesc.shape.count >= 2 && logitsDesc.shape[1] > 0
let logitsRef = try GrowingLogitsBuffer(
    device: device, descriptor: descriptor, name: logitsOutputName,
    vocabSize: config.vocabSize,
    maxCapacity: logitsSeqIsStatic ? logitsDesc.shape[1] : config.maxContextLength,
    initialCapacity: logitsSeqIsStatic ? logitsDesc.shape[1] : averageExpectedPromptSize
)
```

With that guard, `COREAI_CHUNK_THRESHOLD=1` (already honored by `ModelConfig.swift:126-142`) runs the multi-token prefill as S=1 steps on such bundles. A possible follow-up is to default the chunk size to 1 when `input_ids` is static S=1, so the env var is not needed; I left that out here to keep the change minimal.

Happy to open a PR with the guard (and a unit test against a static-logits descriptor) if that is easier to take than a patch in an issue.

**Comment by john-rocky** [comment]
Thanks for #204 — a distinct prompt function is exactly what static-shape bundles like these need, and it is good to see it on `main`. It also moved the call site this guard lands on, so here are numbers on current `main` (`de31ba5`, Xcode 27 beta 5 SDK, M4 Max).

The failure still reproduces there, and it is a trap rather than the throw I described above: `CoreAIRuntime/NDArrayDescriptor.swift:136: Fatal error: Shape at dimension 1 of 256 is not a valid substitution for source shape 1`, while preparing the asset. With the same five-line guard applied at the new `prefillLogitsInitialCapacity(...)` site and `COREAI_CHUNK_THRESHOLD=1`, `llm-benchmark -p 128 -g 128 -n 3` runs on two decode-only S=1 bundles (two runs of three trials):

| bundle | prompt tok/s | generation tok/s |
|---|---:|---:|
| Qwen3.5-0.8B int4 (`qwen3_5_0_8b_decode_int4lin`) | 194–224 | 192–221 |
| Qwen3.5-4B int4 (`qwen3_5_4b_decode_int4lin`) | 93–95 | 91–93 |

Prompt speed matches generation speed, as expected when the prompt runs as S=1 steps. The guard alone is not enough: without the env var it gets past construction and traps on the first prompt, because 128 tokens is under the default `chunkThreshold` (1024) and the whole prompt is bound to the static `input_ids` (`… of 128 is not a valid substitution for source shape 1`). That is the follow-up I left out above — keying the chunk size on a static S=1 `input_ids` should remove the need for the env var.

The guard as one commit on top of `de31ba5`: https://github.com/john-rocky/coreai-models/commit/948d5e9f2cc98a790fec5e24b76c9fc4b2b26f68 (branch `issue-212-guard-on-main` on the fork). Happy to open it as a PR with a unit test against a static-logits descriptor, if that is the easier form to take.

Thanks for looking at this.

**Comment by stikves** [comment]
Thank you for the bug report, but this is more of a feature request outside of the scope of the pipelined engine, as it is designed for dynamic inputs.

We do have the static chunked engine for fixed-size inference functions, however looking at the bundle structure, it is probably not a great fit either.

Is there any specific reason the Qwen 3.5 export requires non-batched prefill?

**Comment by john-rocky** [comment]
Thanks for taking a look at the bundle structure — and the feature-request framing is fair, happy to have it treated as one. To your question: the S=1 export is not a preference. Batched prefill needs the recurrent scan in Qwen3.5's linear-attention (GatedDeltaNet) layers, and that scan does not lower on the MPSGraph GPU delegate. In my attempts, S=1 is the only shape that gets these models onto the GPU at all.

The mechanics: the recurrence exports as `torch.ops.higher_order.while_loop`, which the delegate rejects (`'scf.while'` region type mismatch); on the macOS 27 beta the same bundle fails even `cpu_only` (Compiler error 2). At S=1 the scan is a single step, so a loop-free single-step recurrence is numerically identical and removes `scf.while` from the graph entirely. `position_ids` and the KV dimensions stay dynamic, so the bundle classifies as dynamic and the prompt runs as pipelined S=1 steps (`COREAI_CHUNK_THRESHOLD=1`) — hence prompt tok/s ≈ generation tok/s in the numbers above. A chunkwise-parallel (scan-free) prefill formulation would restore batched prefill; I have not solved that yet, and if `'scf.while'` lowers on the delegate at some point, S>1 comes back on its own.

On the static chunked engine, the classification confirms your read: with a single `main` and no `extend_*`/`load_embeddings` functions the bundle is dynamic, so it never reaches that engine. For long prompts I pair these models with a static chunked-prefill companion graph (an unrolled q=16 scan), so the fixed-size case is covered on that side.

If the call is that S=1 static-logits bundles stay outside the pipelined engine's scope, keeping the guard on the fork is no trouble. A one-line note in the docs that the pipelined engine expects a dynamic sequence dimension in the logits output would spare the next person the constructor trap — glad to send that as a docs PR instead, if that is the more useful form. Either way, thanks again for looking at this.

**Comment by stikves** [comment]
Thanks for the detailed analysis.

We have `prefill` graph export incoming with https://github.com/apple/coreai-models/pull/211, which might help with this case.

For the "non-dynamic dynamic engine support", let me take a look. If we can add this trivially without compromising the design, might be a good fallback.

If not, we will have to wait for the underlying mpsgraph fixes.
